### PR TITLE
[TTAHUB-2315] Cache IST/NC on completed training reports

### PR DESCRIPTION
**Pull request by thewatermethod**
## Description of change
- Cache the user name and national center name for completed training reports (add new table + migration)
- Display the national center name in the collaborators dropdown
- Add a script to run to fill in missing cached data for reports already completed (can be removed once run)

## How to test
- Run ```updateCompletedEventReportPilots``` on prod like data and confirm that the script runs and completed training reports
-  Confirm that the collaborators dropdown shows the national center name
- Save a training report and confirm that the new table ```EventReportPilotNationalCenterUsers``` is updated
- View a completed training report and confirm that the information displayed there (as far as user/national center) is correct

## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-2315

### Every PR

- [ ] Meets issue criteria
- [ ] JIRA ticket status updated
- [ ] Code is meaningfully tested
- [ ] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [ ] API Documentation updated
- [ ] Boundary diagram updated
- [ ] Logical Data Model updated
- [ ] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions
- [ ] UI review complete

### Before merge to main

- [ ] OHS demo complete
- [ ] Ready to create production PR

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [ ] Update JIRA ticket status

### CODE REVIEW

File: src/models/hooks/eventReportPilot.js [hunk 1]
```diff
@@ -183,6 +182,58 @@ const updateGoalText = async (sequelize, instance, options) => {
   );
 };
 
+const createOrUpdateNationalCenterUserCacheTable = async (sequelize, instance, options) => {
+  try {
+    const { ownerId, collaboratorIds } = instance;
+    const userIds = [ownerId, ...collaboratorIds].filter((f) => f);
```
Comment (AdamAdHocTeam) [inline src/models/hooks/eventReportPilot.js:188]: can owner be a collab?

File: src/migrations/20240221142726-cache-table-for-training-reports.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
+const { prepMigration, removeTables } = require('../lib/migration');
+
+/** @type {import('sequelize-cli').Migration} */
+module.exports = {
+  async up(queryInterface, Sequelize) {
+    await queryInterface.sequelize.transaction(async (transaction) => {
+      const sessionSig = __filename;
+      await prepMigration(queryInterface, transaction, sessionSig);
+      await queryInterface.createTable('EventReportPilotNationalCenterUsers', {
```
Comment (AdamAdHocTeam) [inline src/migrations/20240221142726-cache-table-for-training-reports.js:9]: Any thought for just making the table NationalCenterUsers?

File: src/models/hooks/eventReportPilot.js (hunk 1 above)
Comment (AdamAdHocTeam) [reply src/models/hooks/eventReportPilot.js:188]: nvm prob doesn't matter how you are using it.

Comment (thewatermethod) [reply src/models/hooks/eventReportPilot.js:188]: yes. we don't need to dedupe though, as we just use them in a query: user.id in (userIds)

File: src/migrations/20240221142726-cache-table-for-training-reports.js (hunk 2 above)
Comment (thewatermethod) [reply src/migrations/20240221142726-cache-table-for-training-reports.js:9]: 😞

File: src/services/event.ts [hunk 3]
```diff
@@ -133,10 +138,32 @@ async function findEventHelper(where, plural = false): Promise<EventShape | Even
     return event;
   }
 
-  let owner: undefined | { id: string; name: string; email: string };
+  let { owner } = event.data;
 
-  if (event.ownerId) {
-    owner = await User.findByPk(event.ownerId, { attributes: ['id', 'name', 'email'], raw: true });
+  if (!owner) {
```
Comment (AdamAdHocTeam) [inline src/services/event.ts:143]: I imagine we would but should we check owner is defiened?

File: src/migrations/20240221142726-cache-table-for-training-reports.js (hunk 2 above)
Comment (AdamAdHocTeam) [reply src/migrations/20240221142726-cache-table-for-training-reports.js:9]: My bad this is a join table

**Review by AdamAdHocTeam — COMMENTED** [review]
Not sure if you can repro this but saving a creator and collab with the NC give the following error:

image

I tried a fresh db also removing and adding common.

**Comment by AdamAdHocTeam** [comment]
> Not sure if you can repro this but saving a creator and collab with the NC give the following error:
> 
> image
> 
> I tried a fresh db also removing and adding common.

The location you suggested fixed it:

image

**Comment by thewatermethod** [comment]
The bug @AdamAdHocTeam found was fixed and this is ready for review

**Review by AdamAdHocTeam — APPROVED** [review]
Looks good the bug is fixed.

I was able to run the script and see data populated for my completed TR's. I was also able to update POC info and see it updated in the table.

File: src/services/event.ts [hunk 4]
```diff
@@ -273,14 +301,29 @@ export async function updateEvent(id: number, request: UpdateEventRequest): Prom
...
+        where: { id: ownerId },
+        attributes: [
+          'id',
+          'nameWithNationalCenters',
+          'email',
+          'name',
+        ],
+        include: [
+          {
+            model: db.NationalCenter,
```
Comment (GarrettEHill) [inline src/services/event.ts:316]: does this need attributes: []

File: src/tools/updateCompletedEventReportPilots.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            userName: collaborator.name,
+            eventReportPilotId: report.id,
+            nationalCenterId: collaborator.nationalCenters[0].id,
+            nationalCenterName: collaborator.nationalCenters[0].name,
+          });
+        });
+      },
+    );
+
+    return db.EventReportPilotNationalCenterUser.bulkCreate(bulkCreate, {
```
Comment (GarrettEHill) [inline src/tools/updateCompletedEventReportPilots.js:67]: please confirm that the id column sequence is updated correctly after using bulkCreate or switch to calling create multiple times

File: src/services/event.ts (hunk 4 above)
Comment (thewatermethod) [reply src/services/event.ts:316]: I don't believe the virtual field works with attributes: []

**Pull request merged**